here-naksha-lib-core/src/main/java/com/here/naksha/lib/core/models/storage/HeapCacheCursor.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2017-2023 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+package com.here.naksha.lib.core.models.storage;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * HeapCacheCursor reads all data (up to provided limit) ahead of time to memory, and that allows to go back and forth.
+ * It's not thread-safe in any way, {@link SeekableCursor} is designed to jump to any position of the cursor, so it doesn't really make sense to use it in parallel.
+ * Use {@link HeapCacheCursor} only when you need to walk through results multiple times, if you need to read data only once - please use {@link ForwardCursor} as it consumes much less memory.
+ *
+ * @param <FEATURE>
+ * @param <CODEC>
+ */
+public class HeapCacheCursor<FEATURE, CODEC extends FeatureCodec<FEATURE, CODEC>>
+    extends SeekableCursor<FEATURE, CODEC> {
+
+  protected static final long BEFORE_FIRST_POSITION = -1;
+
+  protected final ForwardCursor<FEATURE, CODEC> originalCursor;
+
+  protected final Map<Long, ForwardCursor<FEATURE, CODEC>.Row> inMemoryData = new HashMap<>();
```

Comment:
It might be optimized:
- first of all jsons are kept as they are, we can use "intern" FibSet solution
- HashMap works really well for 0 - 100 000 000 rows, if we need to cache more than 1 billion of features we might consider HeapCache solution

Reply:
I do not expect that we get the same JSON back multiple times, but I understand where this comes from, you mean caching across different queries. Very clever and good idea, yes, FibSet would be perfect here, but couldn't this be a general thing for all HeapCacheCursors? Already only read into a FibSet and keep the reference in, ones the client wants to have the real Feature, keep a local copy so that all cursors shared the json and raw geometry (wkb) in a FibSet cache, but eventually individually keep deserialized codecs. I hope I understood this correct, so the FibSet keeps a weak reference to an entity that holds the values from the row (id, uuid, err, json, geometry, ...). When a client really accesses the codec, we create a clone and use the clone?

Reply:
Well, I was thinking about splitting json to save memory, it would cost us a processor time, and probably would not be as efficient as adding just more memory... now when I think about it, it doesn't make sense.

About keeping shared copy between requests - well it's sounds prommising, but I'm affraid that it would cost a lot of memory, it make a lot of sense if we operate on relatively small subset over and over again - but in such case I would say it's client responsibility to cache things as the client has better knowledge how to do it right.

Reply:
Maybe we can later add some global setting to enable global caching, so on-demand? There is plenty of room to improve, for now I would be happen getting the functionality ready.